ring/lifecycler.go

```diff
@@ -763,7 +782,12 @@ func (i *Lifecycler) compareTokens(fromRing Tokens) bool {
 func (i *Lifecycler) autoJoin(ctx context.Context, targetState InstanceState) error {
 	var ringDesc *Desc
 
-	err := i.KVStore.CAS(ctx, i.RingKey, func(in interface{}) (out interface{}, retry bool, err error) {
+	err := i.tokenGenerator.CanJoin(ringDesc.GetIngesters())
```

Comment:
You have to pass `ctx` to the function. Then, inside the function, you have to check the context and return `ctx.Err()` if the context is canceled.